lib/auth/authProvider.js

```diff
@@ -0,0 +1,157 @@
+          errorObject = err;
+        } else if (r.result && r.result['$err']) {
+          errorObject = r.result;
+        } else if (r.result && r.result['errmsg']) {
+          errorObject = r.result;
+        } else {
+          numberOfValidConnections = numberOfValidConnections + 1;
+        }
+
+        // Still authenticating against other connections.
+        if (count !== 0) {
+          return;
+        }
+
+        // We have authenticated all connections
+        if (numberOfValidConnections > 0) {
+          // Store the auth details
+          this.addCredentials(credentials);
+          // Return correct authentication
+          callback(null, true);
+        } else {
+          if (errorObject == null) {
+            errorObject = new MongoError(`failed to authenticate using ${credentials.mechanism}`);
+          }
+          callback(errorObject, false);
+        }
+      });
+    };
+
+    const _execute = _connection => process.nextTick(() => execute(_connection));
+
+    // For each connection we need to authenticate
+    while (connections.length > 0) {
+      _execute(connections.shift());
+    }
+  }
+
+  /**
+   * Adds credentials to store only if it does not exist
+   * @param {MongoCredentials} credentials credentials to add to store
+   */
+  addCredentials(credentials) {
+    let found = false;
+
+    for (let i = 0; i < this.authStore.length; i++) {
+      if (this.authStore[i].equal(credentials)) {
+        found = true;
+        break;
+      }
+    }
+
+    if (!found) {
+      this.authStore.push(credentials);
+    }
+  }
+
+  /**
+   * Re authenticate pool
+   * @method
+   * @param {{Server}|{ReplSet}|{Mongos}} server Topology the authentication method is being called on
+   * @param {Connection[]} connections Connections to authenticate using this authenticator
+   * @param {authResultCallback} callback The callback to return the result from the authentication
+   * @return {object}
+   */
+  reauthenticate(server, connections, callback) {
+    const authStore = this.authStore.slice(0);
+    let count = authStore.length;
+    if (count === 0) {
+      return callback(null, null);
+    }
+
+    for (let i = 0; i < authStore.length; i++) {
+      this._auth(server, connections, authStore[i], function(err) {
+        count = count - 1;
+        if (count === 0) {
+          callback(err, null);
+        }
+      });
+    }
+  }
+
+  /**
+   * Remove authStore credentials
+   * @method
+   * @param {string} source Name of database we are removing authStore details about
+   * @return {object}
+   */
+  logout(source) {
+    this.authStore = this.authStore.filter(credentials => credentials.source !== source);
+  }
+}
+
+/**
+ * This is a result from a authentication strategy
+ *
+ * @callback authResultCallback
+ * @param {error} error An error object. Set to null if no error present
+ * @param {boolean} result The result of the authentication process
+ */
+
+module.exports = { AuthProvider };
```

Comment:
I don't mind exporting this as an object for a single object, but I think we should perhaps be more consistent about this. Right now I believe we tend to export a direct object if there is only one in a file.

Reply:
Yes, but that is not forward looking. If/when we switch to ES Module syntax, this will make things easier.